File: _src/05/Chapter5.VB/Maps/PersonMap.vb
Imports System.Data.Entity.ModelConfiguration
Imports Chapter5.VB.Models

Namespace Maps
    Public Class PersonMap
        Inherits EntityTypeConfiguration(Of Person)
        Public Sub New()
            Me.Property(Function(p) p.FirstName) _
                .HasMaxLength(30) _
                .IsRequired()
            Me.Property(Function(p) p.LastName) _
                .HasMaxLength(30) _
                .IsRequired()
            Me.Property(Function(p) p.MiddleName) _
                .HasMaxLength(1).IsFixedLength().IsUnicode(False)
            Ignore(Function(p) p.FullName)
            Me.Property(Function(p) p.BirthDate) _
                .HasPrecision(1)
            Me.Property(Function(p) p.HeightInFeet) _
                .HasPrecision(4, 2)

            HasMany(Function(p) p.Companies) _
                .WithMany(Function(c) c.Persons) _
                .Map(Sub(m)
                         m.MapLeftKey("PesonId")
                         m.MapRightKey("CompanyId")
                     End Sub)
            Map(Sub(p)
                    p.Properties(Function(m) _
                                        New With {
                                            m.Photo,
                                            m.FamilyPicture})
                    p.ToTable("PersonBlob")
                End Sub)

            Map(Sub(p)
                    p.Properties(Function(m) _
                                    New With {
                                    m.Address,
                                    m.BirthDate,
                                    m.FirstName,
                                    m.HeightInFeet,
                                    m.IsActive,
                                    m.LastName,
                                    m.MiddleName,
                                    m.PersonState,
                                    m.PersonTypeId})
                    p.ToTable("Person")
                End Sub)
        End Sub
    End Class
End Namespace

File: _src/05/Chapter5.CSharp/Maps/PersonMap.cs
using System.Data.Entity.ModelConfiguration;
using Chapter5.CSharp.Models;

namespace Chapter5.CSharp.Maps
{
    public class PersonMap : EntityTypeConfiguration<Person>
    {
        public PersonMap()
        {
            Property(p => p.FirstName)
                .HasMaxLength(30)
                .IsRequired();
            Property(p => p.LastName)
                .IsRequired()
                .HasMaxLength(30);
            Property(p => p.MiddleName)
                .HasMaxLength(1)
                .IsFixedLength()
                .IsUnicode(false);
            Ignore(p => p.FullName);
            Property(p => p.BirthDate)
                .HasPrecision(10);
            Property(p => p.HeightInFeet)
                .HasPrecision(4, 2);
            HasMany(p => p.Phones)
                .WithRequired()
                .HasForeignKey(ph => ph.PersonId);
            HasMany(p => p.Companies)
                .WithMany(c => c.Persons)
                .Map(m =>
                {
                    m.MapLeftKey("PesonId");
                    m.MapRightKey("CompanyId");
                });
            Map(p =>
            {
                p.Properties(ph =>
                    new
                    {
                        ph.Photo,
                        ph.FamilyPicture
                    });
                p.ToTable("PersonBlob");
            });
            Map(p =>
            {
                p.Properties(ph =>
                    new
                    {
                        ph.Address,
                        ph.BirthDate,
                        ph.FirstName,
                        ph.HeightInFeet,
                        ph.IsActive,
                        ph.LastName,
                        ph.MiddleName,
                        ph.PersonState,
                        ph.PersonTypeId
                    });
                p.ToTable("Person");
            });
        }
    }
}
